### ISSUE TITLE
FDS performs worse than vanilla

### ISSUE THREAD

**Issue by jordy101** [YyzHarry/imbalanced-regression#26]
Hi !

Thank you for sharing your work on LDS and FDS. It realy inspired me in my work.

I'm trying to predict price (normalized with Z-score) given features about restaurant and delivery (time, location, distance , etc.). My data is imbalanced and I want my model to perform good whatever the price value is.
To help me understand your work, I re-implemented LDS and FDS (based on your advice). I can see LDS improve my model performance but FDS did worse (see graph below).
![newplot(1)](https://user-images.githubusercontent.com/44983314/174046670-357e9ba2-7a67-44d8-81c4-79e4748bd043.png)

Moreover, when I look how statistics are smoothed, I can see that LDS smooths better than FDS (see graph below).
![newplot(2)](https://user-images.githubusercontent.com/44983314/174046698-08f24f22-4826-4d2c-99db-17ee5256eb9a.png)

Did you ever experienced cases/data where FDS performs worse than vanilla ?

Thank you in advance for your help ! 

My code:
```
# imports
import time

import numpy as np
import torch
import torch.nn.functional as F

from pandas import DataFrame, Series
from scipy.ndimage import convolve1d, gaussian_filter1d
from torch import nn
from torch.utils.data import TensorDataset, DataLoader


# classes
class FDSLayer(nn.Module):
    """

    """
    def __init__(self, input_dim, n_bins: int = 100, kernel_size: int = 5, alpha: float = .9, start_smooth: int = 1):
        super(FDSLayer, self).__init__()
        self.register_buffer('mean', torch.zeros((n_bins, input_dim)))
        self.register_buffer('var', torch.ones((n_bins, input_dim)))
        self.register_buffer('smoothed_mean', torch.zeros((n_bins, input_dim)))
        self.register_buffer('smoothed_var', torch.ones((n_bins, input_dim)))

        self.input_dim = input_dim
        self.alpha = alpha
        self.n_bins = n_bins
        self.kernel_size = kernel_size
        self.half_kernel_size = (kernel_size - 1) // 2
        self.sigma = 2
        self.start_smooth = start_smooth

    def smooth(self, inputs, labels, epoch):
        if epoch < self.start_smooth:
            return inputs
        else:
            labels = labels.squeeze()
            bin_indexes = self._get_bin_indexes(labels)

            factor = torch.clamp(torch.sqrt(self.smoothed_var / self.var), .1, 10)
            return (inputs - self.mean[bin_indexes]) * factor[bin_indexes] + self.smoothed_mean[bin_indexes]

    def _get_bin_indexes(self, labels):
        _, bin_edges = torch.histogram(labels, self.n_bins, density=True)

        return torch.bucketize(labels, bin_edges[1:-1])

    def _get_kernel_window(self):
        base_kernel = [0.] * self.half_kernel_size + [1.] + [0.] * self.half_kernel_size
        kernel_window = gaussian_filter1d(base_kernel, sigma=self.sigma)
        kernel_window = kernel_window / sum(kernel_window)

        return torch.FloatTensor(kernel_window)

    def update_running_stats(self, features, labels):
        labels = labels.squeeze()
        bin_indexes = self._get_bin_indexes(labels)

        new_mean = torch.zeros((self.n_bins, self.input_dim))
        new_var = torch.ones((self.n_bins, self.input_dim))
        for b in torch.unique(bin_indexes):
            a = features[bin_indexes == b]
            if features.size() != 0:
                new_mean[b] = torch.mean(a, dim=0)
                new_var[b] = torch.var(a, dim=0, unbiased=True if a.size(0) != 1 else False)

        self.mean = self.alpha * self.mean + (1 - self.alpha) * new_mean
        self.var = self.alpha * self.var + (1 - self.alpha) * new_var

        fds_kernel_window = self._get_kernel_window()

        smoothed_mean = F.conv1d(F.pad(self.mean.view(1, self.n_bins, self.input_dim).permute(2, 0, 1),
                                       pad=(self.half_kernel_size, self.half_kernel_size),
                                       mode='reflect'),
                                 weight=fds_kernel_window.view(1, 1, -1),
                                 padding=0).permute(2, 0, 1).squeeze()
        self.smoothed_mean = smoothed_mean

        smoothed_var = F.conv1d(F.pad(self.var.view(1, self.n_bins, self.input_dim).permute(2, 0, 1),
                                      pad=(self.half_kernel_size, self.half_kernel_size),
                                      mode='reflect'),
                                weight=fds_kernel_window.view(1, 1, -1),
                                padding=0).permute(2, 0, 1).squeeze()
        self.smoothed_var = smoothed_var


class MLPNetwork(nn.Module):
    def __init__(self, input_dim: int, hidden_units: tuple = (128, 128, 128), lds: bool = False, fds: bool = False,
                 n_bins: int = 20, kernel_size: int = 5, alpha: float = .9, start_smooth: int = 1):
        super(MLPNetwork, self).__init__()

        self.lds = lds
        self.fds = fds
        self.n_bins = n_bins
        self.kernel_size = kernel_size
        self.half_kernel_size = kernel_size // 2
        self.sigma = 2

        input_layer = nn.Linear(input_dim, hidden_units[0])
        self.layers = nn.ModuleList(
            [input_layer] +
            [nn.Linear(hidden_units[i - 1], hidden_units[i]) for i in range(1, len(hidden_units))]
        )
        if self.fds:
            self.fds_layer = FDSLayer(hidden_units[-1], n_bins=n_bins, kernel_size=kernel_size, alpha=alpha,
                                      start_smooth=start_smooth)
        self.output_layer = nn.Linear(hidden_units[-1], 1)

    def forward(self, inputs, labels=None, epoch=None):
        x = inputs
        for layer in self.layers:
            x = torch.relu(layer(x))
        smoothed_features = x.view(x.size(0), -1)
        smoothed_features_ = smoothed_features
        if self.training and self.fds:
            smoothed_features_ = self.fds_layer.smooth(smoothed_features_, labels, epoch)
        x = self.output_layer(smoothed_features_)
        return x, smoothed_features

    def fit(self, inputs, labels, val_inputs, val_labels, epochs: int = 200, batch_size: int = 1024):
        if isinstance(inputs, DataFrame):
            inputs = inputs.values
        if isinstance(val_inputs, DataFrame):
            val_inputs = val_inputs.values
        if isinstance(labels, Series):
            labels = labels.values
        if isinstance(val_labels, Series):
            val_labels = val_labels.values

        # Create train dataloader
        inputs = torch.FloatTensor(inputs)
        labels = torch.FloatTensor(labels)
        if self.lds:
            empirical_label_distribution, bin_edges = torch.histogram(labels, bins=self.n_bins, density=True)
            lds_kernel_window = self._get_kernel_window()
            effective_label_distribution = F.conv1d(F.pad(empirical_label_distribution.view(1, 1, -1),
                                                          pad=(self.half_kernel_size, self.half_kernel_size),
                                                          mode='reflect'),
                                                    weight=lds_kernel_window.view(1, 1, -1),
                                                    padding=0).squeeze()
            weights = 1 / effective_label_distribution[torch.bucketize(labels, bin_edges[1:-1])]
        else:
            weights = torch.ones(labels.size())
        train_dataset = TensorDataset(inputs, labels.view(-1, 1), weights)
        train_dataloader = DataLoader(train_dataset, batch_size, shuffle=True, num_workers=4)

        # Create validation dataloader
        val_inputs = torch.FloatTensor(val_inputs)
        val_labels = torch.FloatTensor(val_labels).view(-1, 1)
        val_dataset = TensorDataset(val_inputs, val_labels)
        val_dataloader = DataLoader(val_dataset, batch_size, shuffle=True, num_workers=4)

        loss_fn = self.weighted_mae_loss
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)

        # Train
        for epoch in range(epochs):
            print(f'Epoch {epoch + 1}/{epochs}')
            t0 = time.time()
            train_loss = 0.

            self.train(True)
            if self.fds:
                latent_feature_record, label_record = [], []
            for i, (train_inputs, train_labels, train_weights) in enumerate(train_dataloader, 1):
                optimizer.zero_grad(set_to_none=True)

                if self.fds:
                    predictions, feat = self(train_inputs, train_labels, epoch)
                    latent_feature_record.extend(feat.data.squeeze().numpy())
                    label_record.extend(train_labels.data.squeeze().numpy())
                else:
                    predictions, _ = self(train_inputs, train_labels)

                loss = loss_fn(predictions, train_labels, train_weights)
                loss.backward()
                optimizer.step()

                train_loss += loss.item()
            self.train(False)

            train_loss = train_loss / i

            val_loss = 0.
            for j, (val_inputs, val_labels) in enumerate(val_dataloader, 1):
                with torch.no_grad():
                    predictions = self(val_inputs, val_labels, epoch)[0]

                    loss = loss_fn(predictions, val_labels)
                    val_loss += loss
            val_loss = val_loss / j

            if self.fds:
                latent_features = torch.from_numpy(np.vstack(latent_feature_record))
                labels_ = torch.from_numpy(np.hstack(label_record))
                self.fds_layer.update_running_stats(latent_features, labels_)

            print(f'time: {time.time() - t0:.1f}s - loss: {train_loss:.4f} - val_loss: {val_loss:.4f}\n')

    def _get_kernel_window(self):
        base_kernel = [0.] * self.half_kernel_size + [1.] + [0.] * self.half_kernel_size
        kernel_window = gaussian_filter1d(base_kernel, sigma=self.sigma)
        kernel_window = kernel_window / sum(kernel_window)

        return torch.FloatTensor(kernel_window)

    def predict(self, inputs):
        if isinstance(inputs, DataFrame):
            inputs = inputs.values

        inputs = torch.FloatTensor(inputs)

        return self(inputs)

    @staticmethod
    def weighted_mae_loss(predictions, targets, weights=None):
        loss = torch.abs(predictions - targets).squeeze()
        if weights is not None:
            loss *= weights.squeeze().expand_as(loss)
        loss = torch.mean(loss)
        return loss

**Comment by YyzHarry** [comment]
Hi - thanks for your interest. Glad to hear that our techniques help.

For your specific dataset, I guess the data you are coping with is tabular data (basically inputs are feature vectors), and a fully connected network is used. Some of my thoughts:
1. Depending on your dataset size (# of training samples) and raw input size (how discriminative the input is), where and how to apply FDS could be different.
2. If the network is easily overfitted to dominating samples, then the inital features learned would be similar across the whole range, after which performing smoothing on such feature space would not help that much. Nevertheless, one can also see what's the case when both LDS+FDS applied.
3. I did not check your code, but you might want to do a sanity check that your FDS is updating the features during training.
4. For a specific dataset, you may need to tune the hyper-parameters of FDS a bit (`kernel`, `ks`, `sigma`), and see if the problem still exists.

Hope these help.
